File: JavaCodeSpace/src/cn/medemede/leecode/H2O.java
package cn.medemede.leecode;

import java.util.concurrent.Semaphore;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;

/**
 * 1117. H2O 生成
 */
public class H2O {
    private final Semaphore h = new Semaphore(2);
    private final Semaphore o = new Semaphore(2);

    public H2O() {

    }

    public void hydrogen(Runnable releaseHydrogen) throws InterruptedException {
        h.acquire();
        // releaseHydrogen.run() outputs "H". Do not change or remove this line.
        releaseHydrogen.run();
        o.release();
    }

    public void oxygen(Runnable releaseOxygen) throws InterruptedException {
        o.acquire(2);
        // releaseOxygen.run() outputs "O". Do not change or remove this line.
        releaseOxygen.run();
        h.release(2);
    }
}

class H2O2 {
    int hNum = 0;
    private ReentrantLock lock = new ReentrantLock();
    Condition oFull = lock.newCondition();
    Condition hFull = lock.newCondition();


    public H2O2() {

    }

    public void hydrogen(Runnable releaseHydrogen) throws InterruptedException {
        lock.lock();
        try {
            while (hNum == 2) {
                hFull.signal();
                oFull.await();
            }
            // releaseHydrogen.run() outputs "H". Do not change or remove this line.
            releaseHydrogen.run();
            hNum++;
            if (hNum == 2) {
                hFull.signal();
            }
        } finally {
            lock.unlock();
        }
    }

    public void oxygen(Runnable releaseOxygen) throws InterruptedException {
        lock.lock();
        try {
            while (hNum <= 0) {
                oFull.signal();
                hFull.await();
            }
            // releaseOxygen.run() outputs "O". Do not change or remove this line.
            releaseOxygen.run();
            hNum -= 2;
            if (hNum <= 0) {
                oFull.signal();
            }
        } finally {
            lock.unlock();
        }
    }
}

File: CppCodeSpace/H2O.cpp
#include <mutex>
#include <condition_variable>

class H2O
{
private:
    int hNum;
    int oNum;
    std::mutex lock;
    std::condition_variable cv;

public:
    H2O()
    {
        hNum = 0;
        oNum = 0;
    }

    void hydrogen(function<void()> releaseHydrogen)
    {
        // 锁是为了保护线程从 "被唤醒 -> wait" 期间不被其他线程打断
        std::unique_lock<std::mutex> l(lock);
        cv.wait(l, [this]()
                { return hNum < 2; });
        // releaseHydrogen() outputs "H". Do not change or remove this line.
        releaseHydrogen();
        hNum++;
        if (hNum == 2 && oNum == 1)
        {
            hNum = 0;
            oNum = 0;
        }
        cv.notify_one();
    }

    void oxygen(function<void()> releaseOxygen)
    {
        std::unique_lock<std::mutex> l(lock);
        cv.wait(l, [this]()
                { return oNum < 1; });
        // releaseOxygen() outputs "O". Do not change or remove this line.
        releaseOxygen();
        oNum++;
        if (hNum + oNum == 3)
        {
            hNum = 0;
            oNum = 0;
        }
        cv.notify_one();
    }
};